From bb332feff08ba80938156668ae71573f8acda3bb Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Fri, 5 May 2023 07:31:14 +0000
Subject: [PATCH 45/53] [Huawei] mm/sharepool: Use mmgrab for add_task

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Use mmgrab to replace mmget in sp_group_add_task. For two reasons:
1. The mmget would protect the mm_struct to release and we have to put
   the usercount in do_exit().
2. The process of sp_group_exit splits into two pieces and there is a
   gap where spg is dead.
Since we pin only the mm_struct instead of the memory space of the
target mm. So we must ensure the existence of the memory space via
mmget_not_zero before we would access it. This is what we did in this
patch.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 30 ++++++++++++++++++++++++++----
 1 file changed, 26 insertions(+), 4 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 43b94002b522..5b946849dddd 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -532,7 +532,6 @@ static int init_local_group(struct mm_struct *mm)
 	if (ret)
 		/* The spm would be released while destroying the spg */
 		goto free_spg;
-	mmget(mm);
 
 	return 0;
 
@@ -543,7 +542,7 @@ static int init_local_group(struct mm_struct *mm)
 	return ret;
 }
 
-/* The caller must hold sp_global_sem */
+/* The caller must hold sp_global_sem and the input @mm cannot be freed */
 static int sp_init_group_master_locked(struct task_struct *tsk, struct mm_struct *mm)
 {
 	int ret;
@@ -1251,6 +1250,12 @@ static int sp_group_link_task(struct mm_struct *mm, struct sp_group *spg,
 	if (ret)
 		goto out_kfree;
 
+	/*
+	 * We pin only the mm_struct instead of the memory space of the target mm.
+	 * So we must ensure the existence of the memory space via mmget_not_zero
+	 * before we would access it.
+	 */
+	mmgrab(mm);
 	master->group_num++;
 	list_add_tail(&node->group_node, &master->group_head);
 
@@ -1422,6 +1427,8 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		}
 	}
 	up_read(&spg->rw_lock);
+	/* We put the mm_struct later to protect the mm from exiting while sp_mmap */
+	mmput(mm);
 
 	return spg_id;
 }
@@ -1798,12 +1805,19 @@ static void sp_munmap(struct mm_struct *mm, unsigned long addr,
 		return;
 	}
 
+	if (unlikely(!mmget_not_zero(mm))) {
+		up_write(&mm->mmap_lock);
+		pr_warn("munmap: target mm is exiting\n");
+		return;
+	}
+
 	err = do_munmap(mm, addr, size, NULL);
 	/* we are not supposed to fail */
 	if (err)
 		pr_err("failed to unmap VA %pK when sp munmap\n", (void *)addr);
 
 	up_write(&mm->mmap_lock);
+	mmput(mm);
 }
 
 /* The caller should hold the write lock for spa->spg->rw_lock */
@@ -2118,7 +2132,13 @@ static int sp_map_spa_to_mm(struct mm_struct *mm, struct sp_area *spa,
 	down_write(&mm->mmap_lock);
 	if (unlikely(mm->core_state)) {
 		up_write(&mm->mmap_lock);
-		pr_info("%s, encountered coredump\n", str);
+		pr_warn("%s, encountered coredump\n", str);
+		return SP_SKIP_ERR;
+	}
+
+	if (unlikely(!mmget_not_zero(mm))) {
+		up_write(&mm->mmap_lock);
+		pr_warn("sp_map: target mm is exiting\n");
 		return SP_SKIP_ERR;
 	}
 
@@ -2126,6 +2146,7 @@ static int sp_map_spa_to_mm(struct mm_struct *mm, struct sp_area *spa,
 	mmap_addr = sp_mmap(mm, spa_file(spa), spa, &populate, prot);
 	if (IS_ERR_VALUE(mmap_addr)) {
 		up_write(&mm->mmap_lock);
+		mmput(mm);
 		pr_err("%s, sp mmap failed %ld\n", str, mmap_addr);
 		return (int)mmap_addr;
 	}
@@ -2146,6 +2167,7 @@ static int sp_map_spa_to_mm(struct mm_struct *mm, struct sp_area *spa,
 		}
 		up_write(&mm->mmap_lock);
 	}
+	mmput(mm);
 
 	return ret;
 }
@@ -3638,7 +3660,7 @@ int sp_group_exit(void)
 	}
 
 	/* match with get_task_mm() in sp_group_add_task() */
-	if (atomic_sub_and_test(master->group_num, &mm->mm_users)) {
+	if (atomic_sub_and_test(master->group_num, &mm->mm_count)) {
 		up_write(&sp_global_sem);
 		WARN(1, "Invalid user counting\n");
 		return 1;
-- 
2.17.1

